### PR TITLE
Implement boosting + show results

### PR DESCRIPTION
**Pull request by VimWizardVersace**

### CODE REVIEW

**Review by erickzul — CHANGES_REQUESTED** [review]
In the file:
```
Invert the normalization on our outputs, and add kriging predictions as
    # constants so the network only predicts the residuals.
    mean_scaler = sp.label_scaler.named_transformers_['mean_std_scaler']
    unscaled_mean = mean_output * mean_scaler.var_ + mean_scaler.mean_
    untransformed_mean = krig_mean + unscaled_mean
    ```
   Don't we also want to inverse transform the kriging feature like we do for the unscaled_mean i.e. unscaled_krig_mean? From what I see we call load_and_scale on the kriging mean and variance columns.

Also please use the rmse calculation from ddf common that's now checked in

File: dnn/variational_inference.ipynb [hunk 1]
```diff
@@ -3,31 +3,47 @@
     {
```
Comment (rothn) [inline dnn/variational_inference.ipynb:3]: <pre class="ql-syntax" spellcheck="false">Line #52.        unscaled_mean = mean_output * mean_scaler.var_ + mean_scaler.mean_
</pre><p>If we're predicting residuals, this should be computed over the residuals instead of the raw values.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tnc-br/ddf-isoscapes/pull/165/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (rothn) [inline dnn/variational_inference.ipynb:3]: <pre class="ql-syntax" spellcheck="false">Line #56.        unscaled_var = var_output * var_scaler.scale_ + var_scaler.min_
</pre><p>If we're predicting residuals, this should be computed over the residuals instead of the raw values.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tnc-br/ddf-isoscapes/pull/165/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (rothn) [inline dnn/variational_inference.ipynb:3]: <pre class="ql-syntax" spellcheck="false">Line #57.        untransformed_var = krig_var + keras.layers.Lambda(lambda t: tf.math.log(1 + tf.exp(t)))(unscaled_var)
</pre><p>Using this technique to predict residuals is tricky... As written, this will force our predicted residuals to only be positive, which means we will only be able to predict <em>higher</em> variance than Kriging produces. I would suggest trying the following:</p><ul><li>Use this instead, which should have the intended effect (Pseudocode, please verify e.g. that you can pass two arguments pairwise to a lambda; otherwise, use the appropriate vector zipping transformation):<code>untransformed_var = keras.layers.Lambda(lambda t,k: tf.math.log(1 + tf.exp(t+k)))(unscaled_var, krig_var)﻿</code></li><li>Manually inspect the data going into the neural network (e.g. "inputs") to make sure it is what you expect, at least once</li><li>[Optional] Compare performance with vs. without variance boosting</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tnc-br/ddf-isoscapes/pull/165/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

**Review by rothn — CHANGES_REQUESTED** [review]
I've posted a review with 2 notes that are likely degrading performance in that network and 1 that is certainly seriously degrading performance. All related to the boosting changes. PTAL

**Review by erickzul — CHANGES_REQUESTED** [review]
In https://screenshot.googleplex.com/6A8LoqXghQA5hxQ
do we want to first put mean before variance? I know you have a test but it doesn't seem like it is catching this iiuc

**Comment by VimWizardVersace** [comment]
> In the file:
> 
> ```
> Invert the normalization on our outputs, and add kriging predictions as
>     # constants so the network only predicts the residuals.
>     mean_scaler = sp.label_scaler.named_transformers_['mean_std_scaler']
>     unscaled_mean = mean_output * mean_scaler.var_ + mean_scaler.mean_
>     untransformed_mean = krig_mean + unscaled_mean
>     ```
>    Don't we also want to inverse transform the kriging feature like we do for the unscaled_mean i.e. unscaled_krig_mean? From what I see we call load_and_scale on the kriging mean and variance columns.
> 
> Also please use the rmse calculation from ddf common that's now checked in
> ```

This isn't necessary if I never applying any sort of scaling to the kriging features, right?

**Comment by rothn** [comment]
> > In the file:
> > ```
> > Invert the normalization on our outputs, and add kriging predictions as
> >     # constants so the network only predicts the residuals.
> >     mean_scaler = sp.label_scaler.named_transformers_['mean_std_scaler']
> >     unscaled_mean = mean_output * mean_scaler.var_ + mean_scaler.mean_
> >     untransformed_mean = krig_mean + unscaled_mean
> >     ```
> >    Don't we also want to inverse transform the kriging feature like we do for the unscaled_mean i.e. unscaled_krig_mean? From what I see we call load_and_scale on the kriging mean and variance columns.
> > 
> > Also please use the rmse calculation from ddf common that's now checked in
> > ```
> 
> This isn't necessary if I never applying any sort of scaling to the kriging features, right?

As we previously discussed, this inverse transform is a potentially helpful thing we can do to help our small model learn better. With the inverse transform, the model only has to output a range in approximately [-3, 3] rather than directly predicting the targets. We never want to apply scaling to the Kriging features, either way, because we want to use properly-scaled values in the KL loss function rather than normalized outputs.

Since we've gone back and forth on this a few times, maybe it would be helpful to cover this in a pair programming session. Would you be open to doing that tomorrow? I'll put something on your calendar.

**Comment by VimWizardVersace** [comment]
> https://screenshot.googleplex.com/6A8LoqXghQA5hxQ

columns_to_keep is just used as an argument to filter columns in the dataframe and doesn't affect the order of columns in the dataframe.

The test in the model code does raise an exception if the columns are out of order. 

I change the order here to make it more intuitive to read.

**Comment by VimWizardVersace** [comment]
> In the file:
> 
> ```
> Invert the normalization on our outputs, and add kriging predictions as
>     # constants so the network only predicts the residuals.
>     mean_scaler = sp.label_scaler.named_transformers_['mean_std_scaler']
>     unscaled_mean = mean_output * mean_scaler.var_ + mean_scaler.mean_
>     untransformed_mean = krig_mean + unscaled_mean
>     ```
>    Don't we also want to inverse transform the kriging feature like we do for the unscaled_mean i.e. unscaled_krig_mean? From what I see we call load_and_scale on the kriging mean and variance columns.
> 
> Also please use the rmse calculation from ddf common that's now checked in
> ```

I can do this, but this would involve a pretty major refactor as I would now need to generate isoscapes as part of train_and_evaluate(). 

Also, some of the models we are experimenting with can not (at least yet) be used to generate isoscapes, so I would still like to fall back on this logic if needed.

**Comment by rothn** [comment]
> 



> > In the file:
> > ```
> > Invert the normalization on our outputs, and add kriging predictions as
> >     # constants so the network only predicts the residuals.
> >     mean_scaler = sp.label_scaler.named_transformers_['mean_std_scaler']
> >     unscaled_mean = mean_output * mean_scaler.var_ + mean_scaler.mean_
> >     untransformed_mean = krig_mean + unscaled_mean
> >     ```
> >    Don't we also want to inverse transform the kriging feature like we do for the unscaled_mean i.e. unscaled_krig_mean? From what I see we call load_and_scale on the kriging mean and variance columns.
> > 
> > Also please use the rmse calculation from ddf common that's now checked in
> > ```
> 
> I can do this, but this would involve a pretty major refactor as I would now need to generate isoscapes as part of train_and_evaluate().
> 
> Also, some of the models we are experimenting with can not (at least yet) be used to generate isoscapes, so I would still like to fall back on this logic if needed.

I do not understand this thread @erickzul

**Comment by VimWizardVersace** [comment]
> > 
> 
> > > In the file:
> > > ```
> > > Invert the normalization on our outputs, and add kriging predictions as
> > >     # constants so the network only predicts the residuals.
> > >     mean_scaler = sp.label_scaler.named_transformers_['mean_std_scaler']
> > >     unscaled_mean = mean_output * mean_scaler.var_ + mean_scaler.mean_
> > >     untransformed_mean = krig_mean + unscaled_mean
> > >     ```
> > >    Don't we also want to inverse transform the kriging feature like we do for the unscaled_mean i.e. unscaled_krig_mean? From what I see we call load_and_scale on the kriging mean and variance columns.
> > > 
> > > Also please use the rmse calculation from ddf common that's now checked in
> > > ```
> > 
> > 
> > I can do this, but this would involve a pretty major refactor as I would now need to generate isoscapes as part of train_and_evaluate().
> > Also, some of the models we are experimenting with can not (at least yet) be used to generate isoscapes, so I would still like to fall back on this logic if needed.
> 
> I do not understand this thread @erickzul

https://docs.google.com/spreadsheets/d/11mN27KPF0OQZ7jEJc_0IUbm9YOiQVsg9cnBD9Ou9DUI/edit#gid=1121609787

Some models use Iso_Oxi_Stack_mean_TERZER and isoscape_fullmodel_d18O_prec_REGRESSION, generated using isoscapes produced by martinelli's lab. It's 1) not clear if we have them 2) if we can continue expecting the lab to produce these isoscapes in the future.

**Comment by erickzul** [comment]
@VimWizardVersace  For handing off the model, we know we can't use those columns from other models and they haven't illustrated yet a path forward to move with them. I would suggest we check in duplicate code to calculate RMSE for something that is not usable by the client. It can be a draft/not-checked in code, but I wouldn't send a PR for it to be in production.

**Review by erickzul — APPROVED** [review]
In line "mean=predicted_value, stdev=predicted_std, n=15)", please change n to 5.
Approved otherwise (but make sure to make this change before merging)

**Comment by rothn** [comment]
I thought the data did not support this approach. Did something change?

**Review by erickzul — CHANGES_REQUESTED** [review]
Blocked because we boosting won't be state of the art.
